src/ghci/mod.rs

```diff
@@ -0,0 +1,360 @@
+//! The core [`Ghci`] session struct.
+
+use std::fmt::Debug;
+use std::fmt::Display;
+use std::process::Stdio;
+use std::sync::atomic::AtomicUsize;
+use std::sync::Arc;
+
+use camino::Utf8PathBuf;
+use itertools::Itertools;
+use miette::IntoDiagnostic;
+use miette::WrapErr;
+use tokio::io::AsyncBufReadExt;
+use tokio::io::BufReader;
+use tokio::process::Child;
+use tokio::process::ChildStderr;
+use tokio::process::Command;
+use tokio::sync::mpsc;
+use tokio::sync::oneshot;
+use tokio::sync::Mutex;
+use tokio::task;
+use tokio::task::JoinHandle;
+use tracing::instrument;
+
+mod stdin;
+use stdin::GhciStdin;
+use stdin::StdinEvent;
+
+mod stdout;
+use stdout::GhciStdout;
+use stdout::StdoutEvent;
+
+mod show_modules;
+use show_modules::ModuleSet;
+
+use crate::buffers::LINE_BUFFER_CAPACITY;
+use crate::event_filter::FileEvent;
+use crate::incremental_reader::IncrementalReader;
+use crate::sync_sentinel::SyncSentinel;
+
+/// The `ghci` prompt we use. Should be unique enough, but maybe we can make it better with Unicode
+/// private-use-area codepoints or something in the future.
+pub const PROMPT: &str = "###~GHCID-NG-PROMPT~###";
+
+/// The name we import `System.IO` as in `ghci`. This is used to run a few `putStrLn` commands and
+/// similar without messing with the user's namespace. If you have a module in your project named
+/// `GHCID_NG_IO_INTERNAL__` that's on you.
+pub const IO_MODULE_NAME: &str = "GHCID_NG_IO_INTERNAL__";
+
+/// A `ghci` session.
+pub struct Ghci {
+    /// A function which returns the command used to start this `ghci` session.
+    /// This needs to be a function because [`Command`] doesn't implement [`Clone`].
```

Comment:
it wasn't clear to me how this relates to `Clone` (i'm still a rust newb)

Reply:
`Clone` makes a copy of a value, but `Command` doesn't want to be copied for whatever reason. But if you put it on the heap with `Box`, you get `Clone` for free because you're copying a pointer to the value instead of the value itself.

@9999years Could this be a `Cow` instead?

Reply:
Actually, we only need the `Box<>` because `fn` is a dynamically-sized type, whereas `Box` is statically-sized (it's exactly the size of a pointer to the heap). The other way to do this is to make `Ghci` generic over a type `F where F: Fn() -> miette::Result<Command> + Send + Sync`, but that would be annoying so we use a `dyn` pointer here (which includes a vtable for dynamic method lookup).

We need to be able to get another copy of the command on demand, so we either need to be able to `Clone` the `Command` (which we can't do) or have a function that provides the `Command` (which is easy and what I chose here).

There's also some kinda-complex interactions with the mechanics of keeping the references to the `ghci` session in `Arc<Mutex<Ghci>>`s... we need the `Command` to construct a new `Ghci` session, but we can't get that `Command` out without leaving invalid memory behind, and we can't get a new valid `Ghci` session without access to the `Command`.

I guess we could replace this with an `Arc<Mutex<Command>>`, but that would also be unideal. Maybe a little more efficient, though.

Reply:
Oh that makes sense, my bad

Reply:
Actually, I do think `Arc<Mutex<Command>>` would be better here, let me change that.